[Bug]: Github resolver looking for wrong PR number

### Is there an existing issue for the same bug?

- [X] I have checked the existing issues.

### Describe the bug and reproduction steps

When responding to an issue on this PR: https://github.com/All-Hands-AI/OpenHands/pull/5058

The github resolver is getting an error like this:
```
ERROR:root:<class 'requests.exceptions.HTTPError'>: 404 Client Error: Not Found for url: https://api.github.com/repos/All-Hands-AI/OpenHands/issues/12809
```

I'm not sure why it's looking for issue 12809 despite the fact that the PR is 5058. We should investigate `openhands-resolver.yml` and any relevant files in the `openhands/resolver` directory to find the error, fix it, and write a test to make sure it doesn't happen again.

### OpenHands Installation

Docker command in README

### OpenHands Version

_No response_

### Operating System

None

### Logs, Errors, Screenshots, and Additional Context

Here is the full trace:

```
Run cd /tmp && python -m openhands.resolver.resolve_issue \
  cd /tmp && python -m openhands.resolver.resolve_issue \
    --repo All-Hands-AI/OpenHands \
    --issue-number 5058 \
    --issue-type pr \
    --max-iterations 50 \
    --comment-id None
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.1[2](https://github.com/All-Hands-AI/OpenHands/actions/runs/11861706839/job/33059571872?pr=5058#step:10:2).7/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/[3](https://github.com/All-Hands-AI/OpenHands/actions/runs/11861706839/job/33059571872?pr=5058#step:10:3).12.7/x64/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.7/x6[4](https://github.com/All-Hands-AI/OpenHands/actions/runs/11861706839/job/33059571872?pr=5058#step:10:4)
    Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.7/x64
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.7/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.12.7/x64/lib
    ISSUE_NUMBER: [5](https://github.com/All-Hands-AI/OpenHands/actions/runs/11861706839/job/33059571872?pr=5058#step:10:5)058
    ISSUE_TYPE: pr
    COMMENT_ID: None
    MAX_ITERATIONS: 50
    SANDBOX_ENV_GITHUB_TOKEN: ***
    GITHUB_TOKEN: ***
    GITHUB_USERNAME: ***
    LLM_MODEL: ***
    LLM_API_KEY: ***
    LLM_BASE_URL: ***
    PYTHONPATH: 
ERROR:root:  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/opt/hostedtoolcache/Python/3.12.7/x[6](https://github.com/All-Hands-AI/OpenHands/actions/runs/11861706839/job/33059571872?pr=5058#step:10:6)4/lib/python3.12/site-packages/openhands/resolver/resolve_issue.py", line 624, in <module>
    main()
  File "/opt/hostedtoolcache/Python/3.12.[7](https://github.com/All-Hands-AI/OpenHands/actions/runs/11861706839/job/33059571872?pr=5058#step:10:7)/x64/lib/python3.12/site-packages/openhands/resolver/resolve_issue.py", line 604, in main
    asyncio.run(
  File "/opt/hostedtoolcache/Python/3.12.7/x64/lib/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.7/x64/lib/python3.12/asyncio/runners.py", line 11[8](https://github.com/All-Hands-AI/OpenHands/actions/runs/11861706839/job/33059571872?pr=5058#step:10:8), in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.7/x64/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.7/x64/lib/python3.12/site-packages/openhands/resolver/resolve_issue.py", line 341, in resolve_issue
    issues: list[GithubIssue] = issue_handler.get_converted_issues(
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.7/x64/lib/python3.12/site-packages/openhands/resolver/issue_definitions.py", line 4[9](https://github.com/All-Hands-AI/OpenHands/actions/runs/11861706839/job/33059571872?pr=5058#step:10:9)7, in get_converted_issues
    closing_issues = self.__get_context_from_external_issues_references(
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.7/x64/lib/python3.12/site-packages/openhands/resolver/issue_definitions.py", line 464, in __get_context_from_external_issues_references
    response.raise_for_status()
  File "/opt/hostedtoolcache/Python/3.12.7/x64/lib/python3.12/site-packages/requests/models.py", line [10](https://github.com/All-Hands-AI/OpenHands/actions/runs/11861706839/job/33059571872?pr=5058#step:10:10)24, in raise_for_status
    raise HTTPError(http_error_msg, response=self)

ERROR:root:<class 'requests.exceptions.HTTPError'>: 404 Client Error: Not Found for url: https://api.github.com/repos/All-Hands-AI/OpenHands/issues/[12](https://github.com/All-Hands-AI/OpenHands/actions/runs/11861706839/job/33059571872?pr=5058#step:10:12)809
Error: Process completed with exit code 1.
```

--- a/openhands/resolver/issue_definitions.py
+++ b/openhands/resolver/issue_definitions.py
@@ -469,17 +469,20 @@ class PRHandler(IssueHandler):
         )
 
         for issue_number in unique_issue_references:
-            url = f'https://api.github.com/repos/{self.owner}/{self.repo}/issues/{issue_number}'
-            headers = {
-                'Authorization': f'Bearer {self.token}',
-                'Accept': 'application/vnd.github.v3+json',
-            }
-            response = requests.get(url, headers=headers)
-            response.raise_for_status()
-            issue_data = response.json()
-            issue_body = issue_data.get('body', '')
-            if issue_body:
-                closing_issues.append(issue_body)
+            try:
+                url = f'https://api.github.com/repos/{self.owner}/{self.repo}/issues/{issue_number}'
+                headers = {
+                    'Authorization': f'Bearer {self.token}',
+                    'Accept': 'application/vnd.github.v3+json',
+                }
+                response = requests.get(url, headers=headers)
+                response.raise_for_status()
+                issue_data = response.json()
+                issue_body = issue_data.get('body', '')
+                if issue_body:
+                    closing_issues.append(issue_body)
+            except requests.exceptions.RequestException as e:
+                logger.warning(f'Failed to fetch issue {issue_number}: {str(e)}')
 
         return closing_issues
 

--- a/openhands/resolver/issue_definitions.py
+++ b/openhands/resolver/issue_definitions.py
@@ -83,7 +83,21 @@ class IssueHandler(IssueHandlerInterface):
         return re.findall(image_pattern, issue_body)
 
     def _extract_issue_references(self, body: str) -> list[int]:
-        pattern = r'#(\d+)'
+        # First, remove code blocks as they may contain false positives
+        body = re.sub(r'```.*?```', '', body, flags=re.DOTALL)
+        
+        # Remove inline code
+        body = re.sub(r'`[^`]*`', '', body)
+        
+        # Remove URLs that contain hash symbols
+        body = re.sub(r'https?://[^\s)]*#\d+[^\s)]*', '', body)
+        
+        # Now extract issue numbers, making sure they're not part of other text
+        # The pattern matches #number that:
+        # 1. Is at the start of text or after whitespace/punctuation
+        # 2. Is followed by whitespace, punctuation, or end of text
+        # 3. Is not part of a URL
+        pattern = r'(?:^|[\s\[({]|[^\w#])#(\d+)(?=[\s,.\])}]|$)'
         return [int(match) for match in re.findall(pattern, body)]
 
     def _get_issue_comments(

--- a/openhands/resolver/issue_definitions.py
+++ b/openhands/resolver/issue_definitions.py
@@ -85,13 +85,13 @@ class IssueHandler(IssueHandlerInterface):
     def _extract_issue_references(self, body: str) -> list[int]:
         # First, remove code blocks as they may contain false positives
         body = re.sub(r'```.*?```', '', body, flags=re.DOTALL)
-        
+
         # Remove inline code
         body = re.sub(r'`[^`]*`', '', body)
-        
+
         # Remove URLs that contain hash symbols
         body = re.sub(r'https?://[^\s)]*#\d+[^\s)]*', '', body)
-        
+
         # Now extract issue numbers, making sure they're not part of other text
         # The pattern matches #number that:
         # 1. Is at the start of text or after whitespace/punctuation

--- a/tests/unit/resolver/test_issue_handler_error_handling.py
+++ b/tests/unit/resolver/test_issue_handler_error_handling.py
@@ -2,12 +2,12 @@ import pytest
 import requests
 from unittest.mock import patch, MagicMock
 
-from openhands.resolver.issue_definitions import IssueHandler
+from openhands.resolver.issue_definitions import PRHandler
 from openhands.resolver.github_issue import ReviewThread
 
 
 def test_handle_nonexistent_issue_reference():
-    handler = IssueHandler("test-owner", "test-repo", "test-token")
+    handler = PRHandler("test-owner", "test-repo", "test-token")
     
     # Mock the requests.get to simulate a 404 error
     mock_response = MagicMock()
@@ -15,7 +15,7 @@ def test_handle_nonexistent_issue_reference():
     
     with patch('requests.get', return_value=mock_response):
         # Call the method with a non-existent issue reference
-        result = handler._IssueHandler__get_context_from_external_issues_references(
+        result = handler._PRHandler__get_context_from_external_issues_references(
             closing_issues=[],
             closing_issue_numbers=[],
             issue_body="This references #999999",  # Non-existent issue
@@ -29,7 +29,7 @@ def test_handle_nonexistent_issue_reference():
 
 
 def test_handle_rate_limit_error():
-    handler = IssueHandler("test-owner", "test-repo", "test-token")
+    handler = PRHandler("test-owner", "test-repo", "test-token")
     
     # Mock the requests.get to simulate a rate limit error
     mock_response = MagicMock()
@@ -39,7 +39,7 @@ def test_handle_rate_limit_error():
     
     with patch('requests.get', return_value=mock_response):
         # Call the method with an issue reference
-        result = handler._IssueHandler__get_context_from_external_issues_references(
+        result = handler._PRHandler__get_context_from_external_issues_references(
             closing_issues=[],
             closing_issue_numbers=[],
             issue_body="This references #123",
@@ -53,12 +53,12 @@ def test_handle_rate_limit_error():
 
 
 def test_handle_network_error():
-    handler = IssueHandler("test-owner", "test-repo", "test-token")
+    handler = PRHandler("test-owner", "test-repo", "test-token")
     
     # Mock the requests.get to simulate a network error
     with patch('requests.get', side_effect=requests.exceptions.ConnectionError("Network Error")):
         # Call the method with an issue reference
-        result = handler._IssueHandler__get_context_from_external_issues_references(
+        result = handler._PRHandler__get_context_from_external_issues_references(
             closing_issues=[],
             closing_issue_numbers=[],
             issue_body="This references #123",
@@ -72,7 +72,7 @@ def test_handle_network_error():
 
 
 def test_successful_issue_reference():
-    handler = IssueHandler("test-owner", "test-repo", "test-token")
+    handler = PRHandler("test-owner", "test-repo", "test-token")
     
     # Mock a successful response
     mock_response = MagicMock()
@@ -81,7 +81,7 @@ def test_successful_issue_reference():
     
     with patch('requests.get', return_value=mock_response):
         # Call the method with an issue reference
-        result = handler._IssueHandler__get_context_from_external_issues_references(
+        result = handler._PRHandler__get_context_from_external_issues_references(
             closing_issues=[],
             closing_issue_numbers=[],
             issue_body="This references #123",

--- a/tests/unit/resolver/test_issue_handler_error_handling.py
+++ b/tests/unit/resolver/test_issue_handler_error_handling.py
@@ -1,0 +1,94 @@
+import pytest
+import requests
+from unittest.mock import patch, MagicMock
+
+from openhands.resolver.issue_definitions import IssueHandler
+from openhands.resolver.github_issue import ReviewThread
+
+
+def test_handle_nonexistent_issue_reference():
+    handler = IssueHandler("test-owner", "test-repo", "test-token")
+    
+    # Mock the requests.get to simulate a 404 error
+    mock_response = MagicMock()
+    mock_response.raise_for_status.side_effect = requests.exceptions.HTTPError("404 Client Error: Not Found")
+    
+    with patch('requests.get', return_value=mock_response):
+        # Call the method with a non-existent issue reference
+        result = handler._IssueHandler__get_context_from_external_issues_references(
+            closing_issues=[],
+            closing_issue_numbers=[],
+            issue_body="This references #999999",  # Non-existent issue
+            review_comments=[],
+            review_threads=[],
+            thread_comments=None
+        )
+        
+        # The method should return an empty list since the referenced issue couldn't be fetched
+        assert result == []
+
+
+def test_handle_rate_limit_error():
+    handler = IssueHandler("test-owner", "test-repo", "test-token")
+    
+    # Mock the requests.get to simulate a rate limit error
+    mock_response = MagicMock()
+    mock_response.raise_for_status.side_effect = requests.exceptions.HTTPError(
+        "403 Client Error: Rate Limit Exceeded"
+    )
+    
+    with patch('requests.get', return_value=mock_response):
+        # Call the method with an issue reference
+        result = handler._IssueHandler__get_context_from_external_issues_references(
+            closing_issues=[],
+            closing_issue_numbers=[],
+            issue_body="This references #123",
+            review_comments=[],
+            review_threads=[],
+            thread_comments=None
+        )
+        
+        # The method should return an empty list since the request was rate limited
+        assert result == []
+
+
+def test_handle_network_error():
+    handler = IssueHandler("test-owner", "test-repo", "test-token")
+    
+    # Mock the requests.get to simulate a network error
+    with patch('requests.get', side_effect=requests.exceptions.ConnectionError("Network Error")):
+        # Call the method with an issue reference
+        result = handler._IssueHandler__get_context_from_external_issues_references(
+            closing_issues=[],
+            closing_issue_numbers=[],
+            issue_body="This references #123",
+            review_comments=[],
+            review_threads=[],
+            thread_comments=None
+        )
+        
+        # The method should return an empty list since the network request failed
+        assert result == []
+
+
+def test_successful_issue_reference():
+    handler = IssueHandler("test-owner", "test-repo", "test-token")
+    
+    # Mock a successful response
+    mock_response = MagicMock()
+    mock_response.raise_for_status.return_value = None
+    mock_response.json.return_value = {"body": "This is the referenced issue body"}
+    
+    with patch('requests.get', return_value=mock_response):
+        # Call the method with an issue reference
+        result = handler._IssueHandler__get_context_from_external_issues_references(
+            closing_issues=[],
+            closing_issue_numbers=[],
+            issue_body="This references #123",
+            review_comments=[],
+            review_threads=[],
+            thread_comments=None
+        )
+        
+        # The method should return a list with the referenced issue body
+        assert result == ["This is the referenced issue body"]

--- a/tests/unit/resolver/test_issue_references.py
+++ b/tests/unit/resolver/test_issue_references.py
@@ -1,0 +1,34 @@
+from openhands.resolver.issue_definitions import IssueHandler
+
+
+def test_extract_issue_references():
+    handler = IssueHandler("test-owner", "test-repo", "test-token")
+
+    # Test basic issue reference
+    assert handler._extract_issue_references("Fixes #123") == [123]
+
+    # Test multiple issue references
+    assert handler._extract_issue_references("Fixes #123, #456") == [123, 456]
+
+    # Test issue references in code blocks should be ignored
+    assert handler._extract_issue_references("""
+    Here's a code block:
+    ```python
+    # This is a comment with #123
+    def func():
+        pass  # Another #456
+    ```
+    But this #789 should be extracted
+    """) == [789]
+
+    # Test issue references in inline code should be ignored
+    assert handler._extract_issue_references("This `#123` should be ignored but #456 should be extracted") == [456]
+
+    # Test issue references in URLs should be ignored
+    assert handler._extract_issue_references("Check http://example.com/#123 but #456 should be extracted") == [456]
+
+    # Test issue references in markdown links should be extracted
+    assert handler._extract_issue_references("[Link to #123](http://example.com) and #456") == [123, 456]
+
+    # Test issue references with text around them
+    assert handler._extract_issue_references("Issue #123 is fixed and #456 is pending") == [123, 456]

